### ISSUE TITLE
[Feature Request] Access to data in Validation rules method

### ISSUE THREAD

**Issue by lucianholt97** [cloudcreativity/laravel-json-api#497]
As part of our validation, we use [custom rules](https://laravel.com/docs/6.x/validation#using-rule-objects) that take other resource attributes as an argument.
By digging into the Validator code, I didn't find any possibility to access the document data which is being validated.
I would love to see this added in the future since I think this is a common use-case.

Thanks in advance!

**Comment by lindyhopchris** [comment]
Yeah, sure. I'm assuming you mean you'd like the data that is going to be validated passed to the `rules()` method on your validators class?

**Comment by lucianholt97** [comment]
Yes, exactly 👍

**Comment by lucianholt97** [comment]
@lindyhopchris are there any updates on this issue?
I would really appreciate it and also found another important use case when validating if the user is authorized to associate a specific relationship.

**Comment by lindyhopchris** [comment]
@lucianholt97 I've had a think about this and I'm not sure about it. Form requests in Laravel don't receive the data being validated to the rules method, as shown here:
https://laravel.com/docs/7.x/validation#form-request-validation

Going forward I'm moving the validation to Laravel form requests, so there won't be the opportunity in the future to have the data passed to the rules method on the form request, as that isn't how Laravel works with form requests.

I find I never need the data for the rules method, because if I need to take into account multiple attributes, using conditional validation is normally the solution, particularly using the `somtimes()` method:
https://laravel.com/docs/7.x/validation#conditionally-adding-rules

For authorization, that logic should be in the authorizer. The authorizer methods provide access to the request, so it's more than possible to write validation logic that is based on something in the JSON body of the request. Again, as we're moving towards form requests - on those there's an `authorize` method and you'd access the JSON body from the form request itself.

So TL;DR - as we're moving to form requests, adding something in that isn't supported by form requests will make the upgrade extremely difficult.

Interested to hear your thoughts.

**Comment by colq2** [comment]
Hello @lindyhopchris , 
I'm with @lucianholt97 in a team and working on this topic.

It's great to hear, that you are moving to Laravel form requests. You are right, that the `rule` function not receives a data object. But as the tip says "You may type-hint any dependencies you need within the rules method's signature. They will automatically be resolved via the Laravel service container."

In order to this you _can_ actually retrieve the data from the request.

```
public function rules(Request $request)
{
    $request->get('something');
}
```

I don't know how much will break, when you move to form request, but I think it should be possible to create a "type hintable" class for getting the data or a helper function.

And yes of course, authorization logic shouldn't be in the validator but in the authorizer.

**Comment by lindyhopchris** [comment]
@colq2 thanks for the reply. There actually wouldn't be any need to type-hint the `Request` class like that on a `FormRequest` - because `FormRequest`s extend `Request`, so you can just do `$this->json()` to access the JSON body content.

The problem with the current `Validators` classes is they are separate from the request, plus the validation data is actually different from the the request body JSON as it is put into a standardised format because in the past I've had people ask that they could just validate `foo` rather than `data.attributes.foo`, or for relationships `bar.id`. instead of `data.relationships.bar.data.id`.

So.... I suppose thinking about it again there isn't really a direct upgrade path from the `Validators` to the `FormRequest` classes, so probably adding the change in #529 probably isn't a problem at this point.

**Comment by lindyhopchris** [comment]
@colq2 I've added some comments to a review of the PR. Those will need resolving before merging.

**Comment by colq2** [comment]
Thank you!

Yes it's way smarter to use `$this` instead of type hint that. I hadn't thought of that. :smile: 

We will look into your comments change the pr.
